File: hash/java/src/main/java/io/rosetta/imagehash/internal/FindSegments.java
package io.rosetta.imagehash.internal;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Pillow-compatible _find_all_segments flood fill.
 *
 * Matches Python imagehash._find_all_segments exactly:
 * 1. Find connected "hill" regions (pixels > threshold), row-major order.
 * 2. Find connected "valley" regions (pixels <= threshold), row-major order.
 * 3. Keep only regions with size > minSegmentSize (STRICT greater-than).
 *
 * Iteration order matches numpy.argwhere row-major (y-first, x-second).
 * Connectivity is 4-neighbor (up/down/left/right).
 *
 * Reference: Python imagehash._find_all_segments (lines 589-634).
 */
public final class FindSegments {

    /**
     * Result segment: ordered list of int[]{y, x} pixel coordinates,
     * sorted in the order they were added (row-major BFS from the first
     * unvisited pixel found in row-major scan).
     */
    public record Segment(List<int[]> pixels) {}

    /**
     * @param pixels         float32 grayscale array [H][W]
     * @param threshold      pixel value threshold; hills = pixels > threshold
     * @param minSegmentSize minimum size (strict >) for a segment to be kept
     * @return list of segments, hills first then valleys
     */
    public static List<Segment> findAllSegments(float[][] pixels,
                                                 float threshold,
                                                 int minSegmentSize) {
        int H = pixels.length;
        int W = pixels[0].length;

        // threshold_pixels[y][x] = pixels[y][x] > threshold
        boolean[][] above = new boolean[H][W];
        for (int y = 0; y < H; y++) {
            for (int x = 0; x < W; x++) {
                above[y][x] = pixels[y][x] > threshold;
            }
        }

        // unassigned_pixels[y][x] = true initially
        boolean[][] unassigned = new boolean[H][W];
        for (int y = 0; y < H; y++) {
            for (int x = 0; x < W; x++) {
                unassigned[y][x] = true;
            }
        }

        // already_segmented: set of encoded (y,x) coords
        // Python seeds with border pixels (-1,z), (H,z), (z,-1), (z,W)
        // We encode interior pixels as y*W+x in [0, H*W)
        // Use a HashSet<Long> encoding: y*bigW + x with out-of-bounds included
        // For efficiency, use a boolean[][] for in-bounds and a HashSet for out-of-bounds
        boolean[][] segmented = new boolean[H][W]; // for in-bounds coords
        // Out-of-bounds border pixels are pre-seeded in Python's already_segmented
        // but they are only needed to prevent _find_region from stepping outside
        // We model this by just using the in-bounds boolean array, and in our
        // BFS we clamp neighbors to valid range while marking them.
        // The key behavior: after processing all hills, the 'while' for valleys
        // checks: len(already_segmented) < H * W
        // already_segmented includes: border pixels (4*(H+W) entries, but
        // in Python it's 2*H + 2*W + 2*H + 2*W = 4*(H+W) border coords)
        // PLUS all pixels that were part of hill segments.
        // The border pixels are OUTSIDE the image (negative coords or >= H/W),
        // so after hills are done, already_segmented.size = borderSize + hillPixelCount
        // borderSize = H + H + W + W = 2*(H+W) in the original Python code
        // (rows: -1..H inclusive = H+2 entries, but let's count exactly)

        // Python border seeding:
        //   (-1, z) for z in range(img_height=W)  -> W entries
        //   (z, -1) for z in range(img_width=H)   -> H entries
        //   (H, z)  for z in range(W)              -> W entries
        //   (z, W)  for z in range(H)              -> H entries
        // Total: 2*W + 2*H entries
        int borderSize = 2 * H + 2 * W;

        List<Segment> segments = new ArrayList<>();

        // --- Pass 1: hill regions (above[y][x] == true) ---
        int hilledCount = 0;  // pixels assigned to hills
        for (;;) {
            // Find first unassigned hill pixel (row-major)
            int startY = -1, startX = -1;
            outer1:
            for (int y = 0; y < H; y++) {
                for (int x = 0; x < W; x++) {
                    if (above[y][x] && unassigned[y][x]) {
                        startY = y;
                        startX = x;
                        break outer1;
                    }
                }
            }
            if (startY < 0) break; // no more hill pixels

            Segment seg = floodFill(above, unassigned, segmented, H, W, startY, startX, true);
            for (int[] p : seg.pixels()) {
                unassigned[p[0]][p[1]] = false;
                hilledCount++;
            }
            if (seg.pixels().size() > minSegmentSize) {
                segments.add(seg);
            }
        }

        // --- Pass 2: valley regions (above[y][x] == false) ---
        // Python: while len(already_segmented) < img_width * img_height
        // already_segmented.size = borderSize + hilledCount + (valley pixels so far)
        // Initially: borderSize + hilledCount pixels are in already_segmented
        // The loop continues while that total < H * W
        // Since borderSize >= 0, if hilledCount >= H*W the loop doesn't run
        int alreadySegmented = borderSize + hilledCount;
        for (;;) {
            if (alreadySegmented >= H * W) break;

            // Find first unassigned valley pixel (row-major)
            int startY = -1, startX = -1;
            outer2:
            for (int y = 0; y < H; y++) {
                for (int x = 0; x < W; x++) {
                    if (!above[y][x] && unassigned[y][x]) {
                        startY = y;
                        startX = x;
                        break outer2;
                    }
                }
            }
            if (startY < 0) break;

            Segment seg = floodFill(above, unassigned, segmented, H, W, startY, startX, false);
            for (int[] p : seg.pixels()) {
                unassigned[p[0]][p[1]] = false;
            }
            alreadySegmented += seg.pixels().size();
            if (seg.pixels().size() > minSegmentSize) {
                segments.add(seg);
            }
        }

        return segments;
    }

    /**
     * BFS flood-fill from (startY, startX) over pixels matching the given polarity.
     * Adds each found pixel to segmented[][].
     *
     * The Python reference uses a set-based BFS (try_next updates via set operations).
     * We replicate row-major iteration order for in_region by sorting the BFS result.
     *
     * Actually, Python's _find_region adds pixels to in_region in a specific order:
     * - 'new_pixels' contains pixels found in this BFS wave
     * - For each pixel in new_pixels, try its 4 neighbors
     * - A neighbor is added to new_pixels (and in_region) if it passes the threshold test
     * The final set 'in_region' is returned; order doesn't matter for the segment itself,
     * but the NEXT region found is determined by row-major scan of unassigned pixels.
     * We just need the SET of pixels; the row-major re-scan happens at the outer loop level.
     */
    private static Segment floodFill(boolean[][] above, boolean[][] unassigned,
                                      boolean[][] segmented, int H, int W,
                                      int startY, int startX, boolean polarity) {
        List<int[]> inRegion = new ArrayList<>();
        Deque<int[]> queue = new ArrayDeque<>();
        // Track what we've seen (either already in segmented OR in this region)
        // We temporarily mark pixels in segmented as we add them to the region
        Set<Long> notInRegion = new HashSet<>();

        int[] start = {startY, startX};
        inRegion.add(start);
        segmented[startY][startX] = true;
        queue.add(start);

        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int cy = cur[0], cx = cur[1];
            for (int[] d : dirs) {
                int ny = cy + d[0], nx = cx + d[1];
                // Out of bounds: these are the border pixels pre-seeded in Python's
                // already_segmented; skip them (they are implicitly "already visited")
                if (ny < 0 || ny >= H || nx < 0 || nx >= W) continue;
                // Already in segmented (hill or previously visited)
                if (segmented[ny][nx]) continue;
                // Already marked as not-in-region by this BFS
                long key = (long) ny * W + nx;
                if (notInRegion.contains(key)) continue;

                if (above[ny][nx] == polarity && unassigned[ny][nx]) {
                    inRegion.add(new int[]{ny, nx});
                    segmented[ny][nx] = true;
                    queue.add(new int[]{ny, nx});
                } else {
                    notInRegion.add(key);
                }
            }
        }

        return new Segment(inRegion);
    }

    private FindSegments() {}
}

File: hash/swift/RosettaImageHash/Sources/RosettaImageHash/Internal/FindSegments.swift
import Foundation

/// Finds all bright ("hill") and dark ("valley") connected segments in a float32 pixel
/// array. Matches Python `imagehash._find_all_segments`.
///
/// Connectivity: 4-neighbor.
/// Iteration order: row-major (y then x), matching `np.argwhere` / `np.nonzero` order.
///
/// The valley loop uses `len(already_segmented) < img_width * img_height` as its
/// termination condition — mirroring the Python source exactly. Note that
/// `img_width` and `img_height` in the Python source map to `pixels.shape[0]` (H)
/// and `pixels.shape[1]` (W) respectively, so `img_width * img_height = H * W`.
///
/// Parameters:
/// - pixels: Float32 array in row-major order (H × W).
/// - width:  W (number of columns).
/// - height: H (number of rows).
/// - threshold: pixels > threshold are "hills". Default 128.
/// - minSegmentSize: minimum pixel count for a surviving segment. Default 500.
///
/// Returns: list of segments; each segment is a list of (y, x) coordinates sorted
/// in row-major order (matching argwhere order).

struct Pixel: Hashable {
    let y: Int
    let x: Int
}

func findAllSegments(
    _ pixels: [Float],
    width: Int,
    height: Int,
    threshold: Float = 128.0,
    minSegmentSize: Int = 500
) -> [[Pixel]] {
    let imgWidth = height  // pixels.shape[0] in Python = H
    let imgHeight = width  // pixels.shape[1] in Python = W

    // Precompute threshold booleans
    var threshPixels = [Bool](repeating: false, count: height * width)
    for i in 0..<(height * width) {
        threshPixels[i] = pixels[i] > threshold
    }

    var unassigned = [Bool](repeating: true, count: height * width)

    var segments: [[Pixel]] = []

    // Already-segmented set: uses a flat representation for image pixels,
    // plus a separate set for the border pseudo-pixels.
    // We need to track border pixels too since the Python code uses a single set.
    // Border pixels: (-1, z), (H, z), (z, -1), (z, W) — encode as special values.
    // For efficiency, track image pixels in a flat Bool array + border count separately.
    var assignedInImage = [Bool](repeating: false, count: height * width)

    // Count of border pixels: 2*H + 2*W (all unique since corners differ)
    let borderCount = 2 * imgWidth + 2 * imgHeight

    // Assigned count starts with all border pixels
    var assignedCount = borderCount

    func markAssigned(_ y: Int, _ x: Int) {
        let idx = y * width + x
        if !assignedInImage[idx] {
            assignedInImage[idx] = true
            assignedCount += 1
        }
    }

    func isAssigned(_ y: Int, _ x: Int) -> Bool {
        // Out-of-bounds pixels are always "assigned" (border)
        if y < 0 || y >= height || x < 0 || x >= width { return true }
        return assignedInImage[y * width + x]
    }

    // Python's `_find_region` uses two sets — `in_region` and `not_in_region`.
    // `not_in_region` pixels are skipped in future iterations but NOT marked
    // assigned. `findRegionExact` reproduces this exactly.
    func findRegionExact(mask: [Bool]) -> [Pixel] {
        var startY = -1, startX = -1
        outer: for y in 0..<height {
            for x in 0..<width {
                if mask[y * width + x] {
                    startY = y; startX = x
                    break outer
                }
            }
        }
        guard startY >= 0 else { return [] }

        var inRegion = Set<Pixel>()
        var notInRegion = Set<Pixel>()
        var newPixels = Set<Pixel>()

        let start = Pixel(y: startY, x: startX)
        inRegion.insert(start)
        newPixels.insert(start)
        markAssigned(startY, startX)

        while !newPixels.isEmpty {
            var tryNext = Set<Pixel>()
            for p in newPixels {
                for (ny, nx) in [(p.y-1,p.x),(p.y+1,p.x),(p.y,p.x-1),(p.y,p.x+1)] {
                    tryNext.insert(Pixel(y: ny, x: nx))
                }
            }
            // Remove already-segmented (border or image-assigned) and not-in-region
            tryNext = tryNext.filter { p in
                !isAssigned(p.y, p.x) && !notInRegion.contains(p)
            }

            if tryNext.isEmpty { break }

            newPixels = Set<Pixel>()
            for p in tryNext {
                // In-bounds check
                if p.y < 0 || p.y >= height || p.x < 0 || p.x >= width {
                    notInRegion.insert(p)
                    continue
                }
                if mask[p.y * width + p.x] {
                    inRegion.insert(p)
                    newPixels.insert(p)
                    markAssigned(p.y, p.x)
                } else {
                    notInRegion.insert(p)
                }
            }
        }

        // Mark unassigned for pixels in region
        for p in inRegion {
            unassigned[p.y * width + p.x] = false
        }

        return Array(inRegion)
    }

    // ----- Hills -----
    func hillMask() -> [Bool]? {
        var mask = [Bool](repeating: false, count: height * width)
        var any = false
        for y in 0..<height {
            for x in 0..<width {
                let idx = y * width + x
                if threshPixels[idx] && unassigned[idx] {
                    mask[idx] = true
                    any = true
                }
            }
        }
        return any ? mask : nil
    }

    while let mask = hillMask() {
        let seg = findRegionExact(mask: mask)
        if seg.count > minSegmentSize {
            segments.append(seg.sorted { $0.y < $1.y || ($0.y == $1.y && $0.x < $1.x) })
        }
    }

    // ----- Valleys -----
    while assignedCount < imgWidth * imgHeight {
        // Build valley mask: !threshold && unassigned
        var mask = [Bool](repeating: false, count: height * width)
        for y in 0..<height {
            for x in 0..<width {
                let idx = y * width + x
                if !threshPixels[idx] && unassigned[idx] {
                    mask[idx] = true
                }
            }
        }
        let seg = findRegionExact(mask: mask)
        if seg.count > minSegmentSize {
            segments.append(seg.sorted { $0.y < $1.y || ($0.y == $1.y && $0.x < $1.x) })
        }
    }

    return segments
}
